File: Python/basicput.py
"""Example of putting a message to an IBM MQ queue
"""
# -*- coding: utf-8 -*-
# Copyright 2018, 2025 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
import datetime
import logging

import ibmmq as mq

from utils.env import EnvStore

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

def connect():
    """ Establish connection to MQ Queue Manager """
    logger.info('Establishing Connection with MQ Server')
    try:
        cd = None
        if not EnvStore.ccdt_check():
            logger.info('CCDT URL export is not set, will be using json environment client connections settings')

            cd = mq.CD(Version=mq.CMQXC.MQCD_VERSION_11)
            cd.ChannelName = MQDetails[EnvStore.CHANNEL]
            cd.ConnectionName = conn_info
            cd.ChannelType = mq.CMQXC.MQCHT_CLNTCONN
            cd.TransportType = mq.CMQXC.MQXPT_TCP

            logger.info('Checking cipher details')
            # If a cipher is set then set the TLS settings
            if MQDetails[EnvStore.CIPHER]:
                logger.info('Making use of cipher details')
                cd.SSLCipherSpec = MQDetails[EnvStore.CIPHER]

        # Key repository is not specified in CCDT so look in environment settings
        # Create an empty SCO object
        sco = mq.SCO()
        if MQDetails[EnvStore.KEY_REPOSITORY]:
            logger.info('Setting Key repository')
            sco.KeyRepository = MQDetails[EnvStore.KEY_REPOSITORY]

        qmgr = mq.QueueManager(None)

        csp = mq.CSP()
        csp.CSPUserId = credentials[EnvStore.USER]
        csp.CSPPassword = credentials[EnvStore.PASSWORD]

        qmgr.connect_with_options(MQDetails[EnvStore.QMGR],
                                  csp=csp, cd=cd, sco=sco)
        return qmgr

    except mq.MQMIError as e:
        logger.error('Error connecting')
        logger.error(e)
        return None

def get_queue():
    """Establish access to queue"""
    logger.info('Opening queue')
    try:
        # Different ways to open a queue.
        q = mq.Queue(qmgr)

        od = mq.OD()
        od.ObjectName = MQDetails[EnvStore.QUEUE_NAME]
        q.open(od, mq.CMQC.MQOO_OUTPUT)

        return q

    except mq.MQMIError as e:
        logger.error('Error opening queue')
        logger.error(e)
        return None

def put_message():
    """Put message onto queue"""
    logger.info('Attempting put to queue')
    try:
        md = mq.MD()
        md.Format = mq.CMQC.MQFMT_STRING
        msg = str(json.dumps(msg_object))
        queue.put(msg,md)

        logger.info('Put message successful: %s',msg)
    except mq.MQMIError as e:
        logger.error('Error in put to queue')
        logger.error(e)

def build_mq_details():
    """Create the connection details for the queue manager"""
    for key in [EnvStore.QMGR, EnvStore.QUEUE_NAME, EnvStore.CHANNEL, EnvStore.HOST,
                EnvStore.PORT, EnvStore.KEY_REPOSITORY, EnvStore.CIPHER]:
        MQDetails[key] = EnvStore.getenv_value(key)

# Application Logic starts here
logger.info('Application "BasicPut" is starting')

envStore = EnvStore()
envStore.set_env()

MQDetails = {}
credentials = {
    EnvStore.USER: EnvStore.getenv_value(EnvStore.APP_USER),
    EnvStore.PASSWORD: EnvStore.getenv_value(EnvStore.APP_PASSWORD)
}

build_mq_details()
conn_info = EnvStore.get_connection(EnvStore.HOST, EnvStore.PORT)

logger.info('Connection is %s', conn_info)

msg_object = {
    'Greeting': 'Hello from Python! ' + str(datetime.datetime.now())
}

qmgr = None
queue = None

qmgr = connect()
if qmgr is not None:
    queue = get_queue()
if queue is not None:
    put_message()
    queue.close()

if qmgr is not None:
    qmgr.disconnect()

logger.info('Application is ending')

File: Go/src/basicput.go
/**
 * Copyright 2019, 2020 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package main

import (
	"encoding/hex"
	"encoding/json"
	"log"
	"math/rand"
	"github.com/ibm-messaging/mq-golang/v5/ibmmq"
	"mqdevpatterns/src/mqsamputils"
	"os"
	"strings"
	"time"
)

var logger = log.New(os.Stdout, "MQ Put: ", log.LstdFlags)

type message struct {
	Greeting string `json:"greeting"`
	Value    int    `json:"value"`
}

// Main Entry to Put application
// Creates Connection to Queue
func main() {

	logger.Println("Application is Starting")

	logSettings()
	mqsamputils.EnvSettings.LogSettings()

	qMgr, err := mqsamputils.CreateConnection(mqsamputils.FULL_STRING)
	if err != nil {
		logger.Fatalln("Unable to Establish Connection to server")
		os.Exit(1)
	}
	defer qMgr.Disc()

	qObject, err := mqsamputils.OpenQueue(qMgr, mqsamputils.Put)
	if err != nil {
		logger.Fatalln("Unable to Open Queue")
		os.Exit(1)
	}
	defer qObject.Close(0)

	putMessage(qObject)

	logger.Println("Application is Ending")
}

// Output authentication values to verify that they have
// been read from the envrionment settings
func logSettings() {
	logger.Printf("Username is (%s)\n", mqsamputils.EnvSettings.User)
	//logger.Printf("Password is (%s)\n", mqsamputils.EnvSettings.Password)
}

func logError(err error) {
	logger.Println(err)
}

func putMessage(qObject ibmmq.MQObject) {
	logger.Println("Writing Message to Queue")

	// The PUT requires control structures, the Message Descriptor (MQMD)
	// and Put Options (MQPMO). Create those with default values.
	putmqmd := ibmmq.NewMQMD()
	pmo := ibmmq.NewMQPMO()

	// The default options are OK, but it's always
	// a good idea to be explicit about transactional boundaries as
	// not all platforms behave the same way.
	pmo.Options = ibmmq.MQPMO_NO_SYNCPOINT

	// Tell MQ what the message body format is. In this case, a text string
	putmqmd.Format = ibmmq.MQFMT_STRING

	// And create the contents to include a timestamp just to prove when it was created
	msgData := &message{
		Greeting: "Hello from Go at " + time.Now().Format(time.RFC3339),
		Value:    rand.Intn(100)}

	data, err := json.Marshal(msgData)
	if err != nil {
		logger.Println("Unexpected error marhalling data to send")
		logError(err)
		return
	}

	// The message is always sent as bytes, so has to be converted before the PUT.

	// Now put the message to the queue
	logger.Printf("Sending message %s", data)
	err = qObject.Put(putmqmd, pmo, data)

	if err != nil {
		logError(err)
	} else {
		logger.Println("Put message to", strings.TrimSpace(qObject.Name))
		// Print the MsgId so it can be used as a parameter to amqsget
		logger.Println("MsgId:" + hex.EncodeToString(putmqmd.MsgId))
	}
}
